R2D2 v3 Adaptation

ftgoktas opened the pull request:
This PR migrates SWELL from r2d2 v1 to r2d2 v3 API for centralized metadata management.
Updates GetObservations, GetBackground, and SaveObsDiags tasks to use new API syntax. 
Adresses #318

**Key Changes:**
- Updated module loading to use `r2d2-client/sles15_0604` instead of `r2d2/sles15_spack19`
- Modified r2d2 configuration to use v3 API parameters (`data_hub`, `data_store`, `compute_host`)
- Updated SWELL tasks to use new r2d2 v3 API syntax:
  - `GetObservations`: Updated fetch parameters (`item='observation'`, `window_start`, etc.)
  - `GetBackground`: Updated for forecast item fetching
  - `SaveObsDiags`: Updated to store as `item='feedback'` with proper parameters
- Added environment variable handling for r2d2 v3 authentication (`R2D2_USER`, `R2D2_API_KEY`)
- Fixed module loading conflicts between JEDI bundles and r2d2 v3
- Created a script for registering observation, background, and bias correction files with r2d2 v3 schema

**Tested with 3dvar and 3dvar_atmos suites** - successfully fetches observations/backgrounds and stores feedback files.

## Dependencies

- r2d2 v3 infrastructure (nccs-gmao data hub, r2d2-experiments-nccs-gmao data store)
- r2d2-client/sles15_0604 module availability on target platform

ftgoktas committed:
Script to setup new r2d2 credentials

ftgoktas committed:
Create a new swell task to test new r2d2

ftgoktas committed:
Adapt get_observations to new R2D2

ftgoktas committed:
Remove exit()

ftgoktas committed:
Add r2d2 configs (#318)

ftgoktas committed:
Update swell tasks to new R2D2 (#318)

ftgoktas committed:
Update r2d2 version of  save obs diagnostics #318

ftgoktas committed:
Remove unused files #318

ftgoktas committed:
Merge remote-tracking branch 'origin/develop' into feature/r2d2_v3

ftgoktas committed:
Create r2d2 file register script #318

ftgoktas committed:
Add scripts for manual setup for R2D2 #318

ftgoktas committed:
Clean up files (#318)

mranst commented on r2d2_v3_setup.sh:
Are these shell scripts meant to be run externally? I don't see calls to them from within swell

ftgoktas replied:
The individual shell scripts are for loading R2D2 to test without running Swell. Included these under /scripts since they are helpful when testing R2D2 file registration and fetching. Swell itself loads R2D2 in modules and doesn't have to use them.

mranst commented on src/swell/utilities/scripts/ingest_files.py:
We may want to consider adding this functionality as a basic suite within swell, or maybe it could be a command to the main swell frontend

Dooruk replied:
I like the idea of having this functionality as a suite though it won't be trivial (nothing is in the DA world it seems!). Registering files would be only allowed to a certain subgroup of users. Ideally one or two obs. lead from each group/product (atm, ocn, aero, bgc. etc). Most index values will require model specific entries (e.g, horizontal_resolution for background, provider, step etc.). Also we would be using different file formats and so many naming conventions (BUFR, nc4, zarr?).

Anyways, we can bring this up with our meeting with @jeromebarre and @mer-a-o

mer-a-o replied:
Having ingest suite can make storing obs files on R2D2 very convenient. I'd be happy to help with setting it up.

ftgoktas replied:
Created #646  for this.

Dooruk replied:
> Having ingest suite can make storing obs files on R2D2 very convenient. I'd be happy to help with setting it up.

thanks @mer-a-o !

mranst commented on scripts/register_files.py:
```diff
@@ -0,0 +1,259 @@
+
+try:
+    import r2d2
+except ImportError as e:
```
I like the helpful message. You can also handle this by raising an error with a custom message, which provides us with a more consistent way of handling issues
```suggestion
try:
    import r2d2
except ImportError:
    raise ImportError(f'Failed to import r2d2, use \nLoad module: module load r2d2-client/sles15_0604')
```

mranst commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
```diff
@@ -39,13 +39,46 @@ module load other/mepo
+    source ~/.swell_r2d2_credentials.sh
+fi
+
+# Check if required variables are set and provide helpful error if not
```
I would personally prefer handling this kind of thing from within python using `os.environ` (others may disagree). Trying to source the the modules file without the swell credentials file caused me to get exited from discover. This could potentially be a swell wrapper around `r2d2`, or a function within the `TaskBase` class that sets the environment variables. We have a directory that we use frequently for config files under `~/.swell`, you may want to consider putting this file there. Instructions for this file should also be added in docs

ftgoktas replied:
The credentials are now loaded from within Python using `os.environ` in the `load_r2d2_credentials()` function, which is called from `create_r2d2_config()` in the `r2d2` utilities module. The credentials file should be located at `~/.swell/r2d2_credentials.yaml`, and the function safely handles missing files by falling back to existing environment variables, so there's no risk of getting kicked from Discover. I removed all the shell script sourcing from the modules file to avoid those exit issues.

mranst commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
```diff
@@ -39,13 +39,46 @@ module load other/mepo
 module load eva/sles15_milan_1.6.5
 module load jedi_bundle/sles15_skylab9
 
+
```
Is jedi_bundle loading a version of r2d2?

ftgoktas replied:
No, jedi_bundle does not include r2d2. Testing also shows "no module named r2d2" when only jedi_bundle is loaded.

mranst commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
```diff
@@ -57,3 +90,10 @@ ulimit -S -s unlimited
 ulimit -S -v unlimited
 umask 022
 
+#########
```
leftover comments

mranst commented on src/swell/tasks/.save_obs_diags.py.swn:
remove swapfiles

mranst commented on src/swell/tasks/get_background.py:
```diff
@@ -143,21 +143,35 @@ def execute(self) -> None:
-                    experiment=background_experiment)
+                file_extension=file_type.split('.')[-1] if '.' in file_type else 'nc'
+
+                print("\n\n************************************\n")
```
In swell we generally prefer to handle this kind of messaging with `self.logger.info(` so we can potentially filter them with messaging levels

Dooruk replied:
Agreed

mranst commented on src/swell/tasks/get_observation_with_logging.py:
Is this task being used currently?

ftgoktas replied:
Removed

mranst commented on src/swell/tasks/get_observations.py:
```diff
@@ -231,22 +242,26 @@ def execute(self) -> None:
+                        observation_type=observation,
+                        file_extension=bias_file_type.split('.')[-1] if '.' in bias_file_type else bias_file_type,
+                        window_start=background_time.strftime('%Y-%m-%dT%H:%M:%SZ'),
+                        window_length='PT6H',
```
I imagine we don't want the value for `window_length` to be hard-coded, but I'm not sure what the appropriate config value would be, may need to add it

ftgoktas replied:
Will address this here #647

mranst commented on src/swell/tasks/save_obs_diags.py:
```diff
@@ -57,29 +58,78 @@ def execute(self) -> None:
+                )
+                self.logger.info(f'Successfully stored feedback file for {observation}')
+                
+            except Exception as e:
```
If there's a specific error type you're expecting to encounter frequently that's not an issue to the run, I'd try to catch it specifically. We don't necessarily want to let any kind of error slip through

mranst commented on src/swell/tasks/task_questions.py:
```diff
@@ -399,6 +399,22 @@ class TaskQuestions(QuestionContainer, Enum):
 
     # --------------------------------------------------------------------------------------------------
 
+    GetObservationsR2d2 = QuestionList(
```
Looks like this may be left over from a previous task

ftgoktas replied:
Yes, this was for testing. Removing it.

mranst commented on src/swell/tasks/save_obs_diags.py:
```diff
@@ -37,10 +37,11 @@ def execute(self) -> None:
 
         # Get window beginning
-        window_begin = self.da_window_params.window_begin(window_offset)
+        #window_begin = self.da_window_params.window_begin(window_offset) #dto
```
I would reset this section to avoid adding to the git blame unnecessarily

mranst commented on src/swell/tasks/save_obs_diags.py:
```diff
@@ -57,29 +58,78 @@ def execute(self) -> None:
 
-            # Store observation files
-            # -----------------------
+            # Store diagnostic observation files (OUTPUT from RunJediVariationalExecutable)
```
This task is used after other jedi executables as well (hofx, localensembleda, etc)

mranst commented on src/swell/tasks/get_observations.py:
```diff
@@ -154,14 +154,24 @@ def execute(self) -> None:
+                    try:            
+                        r2d2.fetch(**fetch_criteria)
+                        self.logger.info(f"Successfully fetched {target_file}")
+                    except Exceotion as e:
```
```suggestion
                    except Exception as e:
```

mranst commented on r2d2_v3_setup.sh:
```diff
@@ -0,0 +1,24 @@
+export R2D2_HOST=discover-gmao
+export R2D2_COMPILER=intel
+
+source venv_client/bin/activate
```
Need a different way of loading the environment if these need to be in swell

Dooruk replied:
agreed, we should have a .r2d2* file at $HOME to handle this like Michael suggested.

mranst reviewed (dismissed):
Nice work- It seems like r2d2 v3 requires a lot of environment variables to be set in order to run. @ashiklom or @Dooruk may have a better idea, but maybe a way to handle this would be to have a config file called `~/.swell/r2d2_credentials.yaml` where the user could store this information. There could then be a function within `TaskBase` that parses and sets these variables using `os.environ` (you can see a somewhat similar example to this in `slurm.py`, where we set slurm defaults in `~/.swell/swell-slurm.yaml`). We're also going to need some documentation explaining this, I'm having trouble getting it running with all the necessary environment additions

Dooruk commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
```diff
@@ -39,13 +39,46 @@ module load other/mepo
-module load r2d2/sles15_spack19
+# Ensure to unload old r2d2 v2 modules
+module unload r2d2 2>/dev/null || true
+module load r2d2-client/sles15_0604
```
So I see two different r2d2 installations, r2d2 and r2d2-client, are both of them necessary?

Since R2D2 uses spack-stack modules, the overall approach shouldn't change, other than setting environment varialbes or API key.

Dooruk commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
same change needs to happen on nccs_discover_cascade

Dooruk reviewed (changes requested):
@ftgoktas great job! finally v3 is becoming a reality.

There are Swell specific peculiarities, @mranst pointed out most of them. Also there are built-in functions like logger so you could use them. We should store R2D2 related config and environment variables inside a file like `~/.swell/r2d2-config.yaml`.

 A few additional notes and Q.s:
 
- There is a little bit of separation happening here with the atmosphere side accessing obs file outside of R2D2 context, which is fine during the development.
 
- Once we have something like `~/.swell/r2d2-config.yaml` we need to have a dedicated section in the documentation, maybe pointing at JCSDA-internal/R2D2 documentation and with useful commands.

- Is it possible to request API keys for @mranst and @ashiklom? Or is it possible to request a generic "gmao-user" API key and have special ones for sudo type users? That way we can create a generic `~/.swell/r2d2-config.yaml` if none exists, otherwise I imagine %98 users will be flabbergasted and message one of us without looking at the documentation.

- I probably asked this before but Is linking rather than copying an option with `fetch`?

- (Long term) Let's keep in mind we would like to use R2D2 on AWS or NAS and able to access cloud dataset (e.g., NNJA) with it.

ftgoktas committed:
Clean up the files (#318)

ftgoktas committed:
Update Python coding norms (#318)

ftgoktas committed:
Fix pycode styles

Dooruk commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
```diff
@@ -39,13 +39,46 @@ module load other/mepo
 module use -a /discover/nobackup/projects/gmao/advda/JediOpt/modulefiles/core
 module load solo/sles15_skylab9
```
Need to eliminate py-boto3 and solo

ftgoktas committed:
Remove redundant lines

ftgoktas committed:
Load R2D2 credentials under TaskBase (#318)

Dooruk commented on r2d2_credentials.yaml:
```diff
@@ -0,0 +1,13 @@
+api_key: your_api_key_here
+
+# Platform configuration  
+host: discover-gmao
```
This should be platform specific (which there is already a way to do this in `utilities/r2d2.py`)

ftgoktas committed:
Load credentials under create R2D2 config (#318)

ftgoktas committed:
make R2D2 host/compiler detection support dynamic (#318)

ftgoktas committed:
Add docs for credential setup (#318)

Dooruk commented on src/swell/tasks/get_observation_with_logging.py:
do we need this file?

Dooruk commented on src/swell/deployment/platforms/nccs_discover_sles15/r2d2_config.yaml:
Make the same change in `src/swell/deployment/platforms/nccs_discover_cascade/r2d2_config.yaml`

ftgoktas committed:
Update r2d2_config for cascade (#318)

ftgoktas committed:
Move credentials under create_task (#318)

Dooruk commented:
Well, `3dvar` is working for me, I was able to use my API key. Finally, R2D2v3 is here, this is exciting 🤩 

One minor comment. If you look at the log output of `GetObservations` task, it is less "readable":

For instance, mine is here:
`~/cylc-run/new_r2d2-suite/runN/log/job/20210701T1200Z/GetObservations-geos_marine/NN/job.out`


<img width="670" height="323" alt="Screenshot 2025-09-24 at 4 13 33 PM" src="https://github.com/user-attachments/assets/f7541f19-84af-44f4-a26b-ff9c800072d1" />

Currently it looks more readable like so:


<img width="1163" height="385" alt="Screenshot 2025-09-24 at 4 16 07 PM" src="https://github.com/user-attachments/assets/a54f0f05-d676-459d-a49c-9f63d065003c" />


Could it be because we are calling r2d2.fetch rather than:

```python
from r2d2 import fetch



fetch(...)
```
?

ftgoktas committed:
Move scripts under utilities (#318)

ftgoktas committed:
Fix pylint errors

ftgoktas commented:
> Well, `3dvar` is working for me, I was able to use my API key. Finally, R2D2v3 is here, this is exciting 🤩
> 
> One minor comment. If you look at the log output of `GetObservations` task, it is less "readable":
> 
> For instance, mine is here: `~/cylc-run/new_r2d2-suite/runN/log/job/20210701T1200Z/GetObservations-geos_marine/NN/job.out`
> 
> <img alt="Screenshot 2025-09-24 at 4 13 33 PM" width="670" height="323" src="https://private-user-images.githubusercontent.com/38666458/493561837-f7541f19-84af-44f4-a26b-ff9c800072d1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg4MjkzODUsIm5iZiI6MTc1ODgyOTA4NSwicGF0aCI6Ii8zODY2NjQ1OC80OTM1NjE4MzctZjc1NDFmMTktODRhZi00NGY0LWEyNmItZmY5YzgwMDA3MmQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI1VDE5MzgwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI0ZWY5YzIwNmYyNDYxYjcwMjZhZGQ1NDU5MDgyMmRiZDJmNGVmNDU1NWUyZTRlMDJiY2M1MTAyNDYzNjQ4YjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fYvEPHh0h34R8purEGpwl2g9iTpSNqhJav833AvYos4">
> Currently it looks more readable like so:
> 
> <img alt="Screenshot 2025-09-24 at 4 16 07 PM" width="1163" height="385" src="https://private-user-images.githubusercontent.com/38666458/493562993-a54f0f05-d676-459d-a49c-9f63d065003c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg4MjkzODUsIm5iZiI6MTc1ODgyOTA4NSwicGF0aCI6Ii8zODY2NjQ1OC80OTM1NjI5OTMtYTU0ZjBmMDUtZDY3Ni00NTlkLWE0OWMtOWY2M2QwNjUwMDNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI1VDE5MzgwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2Y2MxODk5Y2IxM2FlMTBlOWEyZjkwMDk1NWIwZDU5Zjc2Y2YxODlhNjgwMGExOTUzNDZjY2ZmYWM4OGUxMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Mn2Lr1cktiu7hAZ-t4K_nI-p_ltt2iRZxJvgIGFHCys">
> Could it be because we are calling r2d2.fetch rather than:
> 
> ```python
> from r2d2 import fetch
> 
> 
> 
> fetch(...)
> ```
> 
> ?

No, the import style (`import r2d2` vs `from r2d2 import fetch`) doesn't affect the logging output. The difference in logs is because R2D2 v3 has different internal logging compared to the older version, v3's logging is more technical by default.   We can control the logging by setting the `R2D2_LOG_LEVEL` environment variable to `SILENT`, `INFO`, `DEBUG`, or `TRACE` (default is `INFO`). Setting it to `SILENT` will reduce the technical output but to get back the readable format like the old version, we need to add our own custom logging before the `r2d2.fetch()` call.

For comparison:

`R2D2_LOG_LEVEL = “SILENT”`:
￼
<img width="825" height="518" alt="Screenshot 2025-09-25 at 3 20 53 PM" src="https://github.com/user-attachments/assets/6841b78d-29cf-4c40-bd75-3320ccf59ef2" />

`R2D2_LOG_LEVEL = “INFO”` (this is default):
￼
<img width="824" height="647" alt="Screenshot 2025-09-25 at 3 27 07 PM" src="https://github.com/user-attachments/assets/ff66c360-2c32-417d-9d02-2cd9921c71a7" />

`R2D2_LOG_LEVEL = “DEBUG”`:
￼
<img width="822" height="653" alt="Screenshot 2025-09-25 at 3 29 59 PM" src="https://github.com/user-attachments/assets/3ee5d813-d7c5-4b80-9a5e-03b7f490b7f6" />

`R2D2_LOG_LEVEL = “TRACE”`:

<img width="830" height="651" alt="following" src="https://github.com/user-attachments/assets/4baf211a-396c-4c62-b0bf-d0ccd730faa7" />

mranst commented:
3dvar works for me too, get_obs for 3dvar_atmos fails with this error:
<img width="2822" height="824" alt="image" src="https://github.com/user-attachments/assets/18364cb7-0f53-4b42-ba05-6b56e9a3dd69" />

ftgoktas committed:
Merge branch 'develop' of https://github.com/GEOS-ESM/swell into feature/r2d2_v3

ftgoktas committed:
Merge branch 'feature/r2d2_v3' of https://github.com/GEOS-ESM/swell into feature/r2d2_v3

ftgoktas committed:
Fix AttributeError when fetching bias correction files (#318)

ftgoktas committed:
Fix bias correction arguments (#318)

ftgoktas committed:
Fix bias correction argument (#318)

ftgoktas committed:
Add file type argument (#318)

ftgoktas committed:
Fix bias correction ingest

ftgoktas committed:
Improve file extension support

ftgoktas committed:
Add logging

ftgoktas committed:
Resolve merge conflict

ftgoktas committed:
Fix bias coefficient ingestion

ftgoktas committed:
Go back to existing bias naming convention

ftgoktas committed:
Use JCSDA enums for bias files

ftgoktas committed:
Improve logging

ftgoktas commented:
> 3dvar works for me too, get_obs for 3dvar_atmos fails with this error: <img alt="image" width="2000" height="824" src="https://private-user-images.githubusercontent.com/178604374/499008394-18364cb7-0f53-4b42-ba05-6b56e9a3dd69.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjEzNTA4NDMsIm5iZiI6MTc2MTM1MDU0MywicGF0aCI6Ii8xNzg2MDQzNzQvNDk5MDA4Mzk0LTE4MzY0Y2I3LTBmNTMtNGI0Mi1iYTA1LTZiNTZlOWEzZGQ2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyNVQwMDAyMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05YWNjZWVjNDhlZGQxMjRhMmE4ODI5NDRjZjVhMzgwNjExMTA4OWI4ZWQ5OWIyNjhhMjI0M2IxODk1NGEwZjkxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.SeqX1WbGx47d-Jp_VG_dc41fRJfttZiZQthQzDVYs-s">

All 3dvar atmospheric bias coefficient files have been ingested to R2D2. It should be able to run without any errors now.

ftgoktas committed:
Fix code style

ftgoktas committed:
extend exclude list to ignore venv and build directories

Dooruk commented:
Is this still a draft or ready for testing?

ftgoktas commented:
Can be tested.

ftgoktas committed:
Change the script name register_files with ingest_files

Dooruk commented:
Ok, so far I received two errors:

1) This one seems more critical and not sure why am I the only one receiving it. This is what I got while running `3dvar_atmos`:
<img width="1434" height="248" alt="Screenshot 2025-10-27 at 2 04 04 PM" src="https://github.com/user-attachments/assets/2cc4f814-24fd-44d2-96ba-dd6eeaa86ea5" />

2) This one is a simple fix which is a problem in line 287 in `get_observations.py`, I received it during my `3dvar` run. IMO we don't need the self.logger.info outputs there anyway as r2d2 puts out relevant information:

<img width="1021" height="109" alt="Screenshot 2025-10-27 at 2 06 23 PM" src="https://github.com/user-attachments/assets/66b6e88d-fe8e-4d50-904f-60915fc11761" />

mranst commented:
> Ok, so far I received two errors:
> 
>     1. This one seems more critical and not sure why am I the only one receiving it. This is what I got while running `3dvar_atmos`:
> 
> 
> <img alt="Screenshot 2025-10-27 at 2 04 04 PM" width="1434" height="248" src="https://private-user-images.githubusercontent.com/38666458/506162262-2cc4f814-24fd-44d2-96ba-dd6eeaa86ea5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE1OTI2MDQsIm5iZiI6MTc2MTU5MjMwNCwicGF0aCI6Ii8zODY2NjQ1OC81MDYxNjIyNjItMmNjNGY4MTQtMjRmZC00NGQyLTk2YmEtZGQ2ZWVhYTg2ZWE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDI3VDE5MTE0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlNWQ5Mjg2NDIyZmNiMmM2MjEwNTg1ZDI3NDQ2NTQzYTgyZmYxMDZkMThkNzhjMjU2NTI1ZGE2MGU0ODI5ZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.g7f5Ht9D63EidAVqitTGP6Sz7HPIC7VXNQgsrHv5Hj0">


I experienced this as well, maybe a credential issue?

ftgoktas committed:
Fix logger

ftgoktas committed:
Fix string literal

ftgoktas commented:
Fixed read permissions on the bias_correction data store via gmaoswell credentials. Also updated logging. 
It should run without errors now. 

<img width="467" height="429" alt="Screenshot 2025-10-27 at 4 39 47 PM" src="https://github.com/user-attachments/assets/6dd1509c-d395-4d09-9532-c0fdbd6fbf10" />

ftgoktas committed:
Merge remote-tracking branch 'origin/feature/r2d2_v3' into feature/r2d2_v3

ftgoktas committed:
Moved readme to docs

mer-a-o commented on register_swell_3dvar_experiment.py:
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
```
Why there's a need to register only `swell-3dvar` experiment? and why for all the `lifetime` categories? 
IMO this code needs to live under `utilities` and can be made more generic so it can be used to register different experiment names. 

The way it's used in Skylab/Ewok is that experiment `name` is randomly generated when an experiment is created and then the experiment is [registered on R2D2 ](https://github.com/JCSDA-internal/ewok/blob/develop/src/ewok/bin/create_experiment.py#L39)with some [metadata](https://github.com/JCSDA-internal/ewok/blob/develop/src/ewok/r2d2_util.py#L38-L44). By default the `lifetime` is set to `debug` and the experiment (and the associated files) are removed from R2D2 after 2 weeks. 

In SWELL, since we are reusing the same experiment names (and same gmao users) we can register each experiment once with `release` lifetime.

ftgoktas replied:
This file should already be removed in the latest version; could you check if your branch is up to date?

mer-a-o commented on src/swell/tasks/get_observations.py:
```diff
@@ -154,14 +166,23 @@ def execute(self) -> None:
+                        'target_file': target_file,          # Where to save
+                    }
+
+                    print(f"Searching for file with criteria: {fetch_criteria}")
```
Does R2D2 already print this information and handle the error?

ftgoktas replied:
Correct, removing this.

ftgoktas committed:
Fix pycode style

mer-a-o commented on src/swell/tasks/save_obs_diags.py:
```diff
@@ -57,29 +57,78 @@ def execute(self) -> None:
+                else:
+                    self.logger.info(f'  File does not exist!')
+
+                self.logger.info(f'  GetObservations task did not fetch this file successfully')
```
There could be other reasons for the feedback (obs_diag) file to not exist. I suggest removing this line.

mer-a-o commented on src/swell/tasks/save_obs_diags.py:
```diff
@@ -57,29 +57,78 @@ def execute(self) -> None:
+
+            self.logger.info(f'Checking observation {observation}: use_obs = {use_obs}')
+
             if not use_obs:
```
For my understanding, when is this if statement true? Don't we do this check in line 64 with `check_obs`?

ftgoktas replied:
check_obs returns False (so the if statement is True) when the input obs isn’t usable for this cycle. It can be a missing/empty file, a zero location dimension etc.

mer-a-o replied:
Thanks @ftgoktas. Would this make the if statement in line 72 redudnet? You already check the file dimension with `check_obs`.

Dooruk replied:
I agree with @mer-a-o here, I think we are repeating what `check_obs` does here.

ftgoktas replied:
Agreed, removing it.

ftgoktas committed:
Remove redundant lines

ftgoktas committed:
Make the comment more generic

ftgoktas committed:
Merge branch 'develop' into feature/r2d2_v3

ftgoktas committed:
Add readme for ingest_files.py (#318)

ftgoktas committed:
Merge remote-tracking branch 'origin' into feature/r2d2_v3

ftgoktas committed:
Merge remote-tracking branch 'origin/feature/r2d2_v3' into feature/r2d2_v3

Dooruk commented on src/swell/deployment/platforms/nccs_discover_cascade/r2d2_config.yaml:
```diff
@@ -1,22 +1,21 @@
 
+# Cache settings
 cache_name: ${USER}
+cache_fetch: false
```
What is the purpose of `cache_fetch` (just wondering)

ftgoktas replied:
I don't think r2d2_config.yaml is relevant for v3, we can remove it. Only the credential loading is needed for v3.

Dooruk commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
```diff
@@ -37,15 +37,28 @@ module load other/mepo
+
+# Remove any r2d2 v2 paths from PYTHONPATH to prevent conflicts
+export PYTHONPATH=$(echo $PYTHONPATH | tr ':' '\n' | grep -v "/r2d2/sles15_spack19/" | tr '\n' ':' | sed 's/:$//')
+
```
```suggestion

```

I really don't think these are necessary.

Dooruk commented on src/swell/deployment/platforms/nccs_discover_sles15/modules:
```diff
@@ -37,15 +37,28 @@ module load other/mepo
-module load r2d2/sles15_spack19
+
+# Ensure to unload old r2d2 v2 modules
+module unload r2d2 2>/dev/null || true
```
```suggestion

```

I really don't think these are necessary.

Dooruk commented on src/swell/deployment/platforms/nccs_discover_sles15/r2d2_config.yaml:
```diff
@@ -1,22 +1,23 @@
-databases:
+# TODO
 
```
```suggestion
```
you already did it :)

Dooruk commented on src/swell/tasks/base/task_base.py:
```diff
@@ -278,11 +278,22 @@ def create_task(
         ensemblePacket: Optional[str]
     ) -> taskBase:
 
+        # Load R2D2 credentials before importing any task modules
+        # -------------------------------------------------------
+        from swell.utilities.config import Config
+        from swell.utilities.r2d2 import load_r2d2_credentials
+        from swell.utilities.logger import get_logger
+
+        # Get platform info from config to load credentials
+        temp_logger = get_logger('R2D2Setup')
+        temp_config = Config(config, temp_logger, task, model)
+        load_r2d2_credentials(temp_logger, temp_config.__platform__)
```
Could you remind me why this part has to be in the base class and not handled only in R2D2 specific tasks?

ftgoktas replied:
If I ran this inside an R2D2-specific task, I tried 🙃, that module would be imported first, `import r2d2` would run at module load, and credentials wouldn’t be set yet and causes initialization errors.
Putting it in the base class guarantees credentials are loaded before any task module. Will take a look at this again to see if it can be any other way.

Dooruk replied:
I see.. Ok, this is not a show stopper so we can also address it later. Just need to make a note if it or create a minor issue.

Dooruk commented on src/swell/tasks/get_background.py:
```diff
@@ -13,14 +13,13 @@
 r2d2_model_dict = {
     'geos_atmosphere': 'geos',
-    'geos_marine': 'mom6_cice6_UFS',
+    'geos_marine': 'mom6',  # 'mom6_cice6_UFS'
```
One thing for marine is that we deal with different types of models, such as MOM6, CICE6, or WAVEWATCH3 ( even a biogeochemical model) in the future. So perhaps this could be a tuple and depending on the `marine_models` key in `experiment.yaml` only relevant models would be called.

Dooruk replied:
`geos_marine`: ['mom6', 'cice6', 'ww3']

Dooruk commented on src/swell/tasks/get_observations.py:
```diff
@@ -102,6 +108,10 @@ def execute(self) -> None:
 
+        # Get model component
+        model_component = self.get_model()
+        r2d2_model = r2d2_model_dict.get(model_component, model_component)
```
Is this only relevant for bias coeff and satellite time lapse type files?

Dooruk commented on src/swell/tasks/get_observations.py:
```diff
@@ -205,49 +221,87 @@ def execute(self) -> None:
+                    self.logger.info(f'fetching file_extension: {bias_file_ext}')
+                    self.logger.info(f'fetching file_type: {bias_file_type}')
+                    self.logger.info(f'fetching date: {background_time_iso}')
+                    self.logger.info(f'fetching target_file: {target_bccoef}')
```
```suggestion

```

redundant, r2d2 outputs these

Dooruk commented on src/swell/tasks/get_observations.py:
```diff
@@ -205,49 +221,87 @@ def execute(self) -> None:
+                    # 2023-10-09T15:00:00Z
+                    self.logger.info(f'fetching date: {background_time_iso}')
+                    # /path/to/bias_coef.nc
+                    self.logger.info(f'fetching target_file: {target_bccoef}')
```
```suggestion

```
redundant, r2d2 outputs these

Dooruk commented on src/swell/tasks/save_obs_diags.py:
```diff
@@ -9,9 +9,9 @@
-from r2d2 import store
 from swell.utilities.r2d2 import create_r2d2_config
 from swell.utilities.run_jedi_executables import check_obs
+import r2d2
```
nitpick but can you keep `import` statements together 😄 ?

Dooruk commented on src/swell/tasks/save_obs_diags.py:
```diff
@@ -57,29 +57,79 @@ def execute(self) -> None:
+            # ---------------
+            self.logger.info(f'Storing feedback file {obs_path_file} to r2d2')
+            self.logger.info(f'  item=feedback, observation_type={name}')
+            self.logger.info(f'  experiment={self.experiment_id()}')
```
```suggestion
```
redundant, r2d2 outputs these

Dooruk reviewed (changes requested):
I was able to run three suites successfully!

Most of my comments are taking out redundant logger outputs as `r2d2` already puts out that information. For me this is really close, with three things holding this back for me, other than ingesting more obs. ofcourse:

1) There is an addition to the taskBase for r2d2 but could that be put in `utiltiies/r2d2.py` and only employed during R2D2 relevant tasks?
2) Model types should support more for marine (I briefly explained this in my comment)
3) In terms of storage, what "lifetime" are we choosing? Are people storing their experiments temporarily? Is everyone storing their experiments in the general gmao database?


Also, can you bump up the SWELL version? I would like to make a release before we merge this.

I also noticed a slight increase in the time it takes to complete `getObservations`. Not a huge deal as we can trigger this task at a cycle before any other dependencies but just got me thinking.

Dooruk commented:
I probably forgot to mention this one but in cycling tasks, there is a need to save backgrounds and/or restarts. `saveObsDiags.py` is saving observation outputs but we would also like to save model fields, could be forecast, analysis, or restarts. Currently `saveRestart.py` does this in a super manual way but it would be nice R2D2v3 to handle this for us. This is a bit more involved with model runs so I can look into that with the new approach but I need to grasp a few more things in terms of "lifetime".

I am not super clear on where we will end up in terms of `experiment_id` (manual setup vs. randomly assigned) and how long the items are stored within R2D2 but I see there are `lifetime name` values ( debug, science, publication, release). I would say only certain users would have access to `permanent` storage, correct?

@mer-a-o do you store all hourly model state files for the backgrounds? Also do you store ensemble files? (you being in Skylab)

ftgoktas committed:
Update src/swell/deployment/platforms/nccs_discover_sles15/r2d2_config.yaml

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

ftgoktas committed:
Remove redundant loging

Co-authored-by: Doruk Ardağ <38666458+Dooruk@users.noreply.github.com>

ftgoktas committed:
Remove redundant logging

ftgoktas committed:
Remove unnecessary statements

ftgoktas committed:
Bump version to 1.11.0

ftgoktas committed:
Bump version to 1.20.0

ftgoktas committed:
Add how to register experiments to readme

Dooruk commented:
So this fails the `save_restart.py` task, which is basically just a copying operation meant to be used with R2D2v1. However, that lead me to thinking further about a potential issue which I described below:

https://github.com/JCSDA-internal/r2d2/issues/839

For now I can deactivate the `save_restart` and move forward with this PR but I would like to ingest more obs before merging.

@mranst would you mind testing this on `gmao_ci` when you get a chance since this will break tests there unless the API key is setup?

mranst commented:
> @mranst would you mind testing this on `gmao_ci` when you get a chance since this will break tests there unless the API key is setup?

https://github.com/GEOS-ESM/swell/actions/runs/19082156562

I added the `gmao_user` credential file to `gmao_ci`, r2d2 tasks are failing with this error, looks like it's expecting a string true or false and getting a bool type at some point. It's happening for me locally as well

<img width="2790" height="1432" alt="image" src="https://github.com/user-attachments/assets/161b2397-773c-481c-aab7-bd5ea65d1f4a" />

ftgoktas committed:
Upgrade r2d2-client (#618)

ftgoktas commented:
The issue was caused by a bug in the r2d2 codebase, the r2d2-client module on Discover has been updated with the fix, and tasks should now run without errors. @mranst could you please pull the latest changes and re-run the tests?

mranst commented:
> The issue was caused by a bug in the r2d2 codebase, the r2d2-client module on Discover has been updated with the fix, and tasks should now run without errors. @mranst could you please pull the latest changes and re-run the tests?

https://github.com/GEOS-ESM/swell/actions/runs/19150912180/job/54740655821

`SaveRestart` is failing for `3dvar_cycle` and `3dfgat_cycle`, but otherwise looks good

ftgoktas commented:
This will be addressed in #652

ftgoktas committed:
Add register experiment to readme

Dooruk commented:
Working towards merging this, some final things:

- I need to turn off `saveRestart` for this PR. I will create a branch off of yours and make a PR for that.
- `storebackground` is not used currently but we need to revive that (rename it to saveBackground) after this goes in. restart and background are different in DA and modeling context. 

Q.s:
- I see saveobsdiags are in `feedback`. Is `feedback` specific item like a IODA output? i couldn't find much information on this.
- I made a tiny edit to your ingest script and actually testing with that it seems to be working. However there seems to be no protection in terms of who can write into `nccs-gmao` data_hub currently or who can remove files from there, is that correct?

ftgoktas commented:
> Q.s:
> 
> * I see saveobsdiags are in `feedback`. Is `feedback` specific item like a IODA output? i couldn't find much information on this.

Yes, feedback is for IODA observation diagnostic files produced by DA systems, contains original observations plus analysis results. I had confirmed with the R2D2 team that JEDI 3dvar output files (observations + analysis feedback) should use item='feedback', which applies to SaveObsDiags since that task stores those files.

> * I made a tiny edit to your ingest script and actually testing with that it seems to be working. However there seems to be no protection in terms of who can write into `nccs-gmao` data_hub currently or who can remove files from there, is that correct?

The ingest script doesn't enforce permissions; it relies on the R2D2 v3 API server to enforce authentication/authorization, so unauthorized users will get permission errors from the API if they don't have the API key set up in their environment.

Dooruk commented:
Ok, this is ready to go in after #653 merges. There are still things to be addressed but no need to hold this one back.

I made a release right before this so we should be good.

Dooruk committed:
r2d2 v3 turnoffrestart (#653)

Dooruk reviewed (approved):
Excellent job! Long time coming.
